content/posts/type_safe_expressjs_genezio.md

```diff
@@ -0,0 +1,270 @@
+    title: string;
+    author: string;
+}
+
+// genezio: deploy
+export class BookService {
+	books: Book[] = []
+
+	/**
+	* Retrieves all books from the store.
+	* 
+	* @returns {Promise<Book[]>} A promise that resolves with an array of Book objects.
+	*/
+	async getAllBooks(): Promise<Book[]> {
+		return this.books;
+	}
+	
+	/**
+	* Creates a new book and adds it to the store. 
+	* @param {string} title - The title of the book. 
+	* @param {string} author - The author of the book.
+	* @returns {Promise<Book>} A promise that resolves with the newly created Book object.
+	*/
+	async createBook(title: string, author: string): Promise<Book> {
+	    if (!title || !author) {
+            throw new Error('Title and author are required.');
+        }
+
+		const newBook: Book = {
+		   id: this.books.length + 1, // Simple ID generation strategy
+		   title,
+		   author,
+		};
+
+		// Adding the new book to our in-memory store
+		this.books.push(newBook);
+
+		return newBook;
+	}
+}
+
+```
+
+We have to include this class in express using an adapter to make it available and waiting for requests on port "8881". Install the `@genezio/adapters` package in the server folder.
+This adapter will take genezio classes and it will expose them publicly with express.
+
+```bash
+npm install @genezio/adapters
+```
+
+```ts
+import  express from 'express';
+import * as genezioAdapters from "@genezio/adapters";
+import { BookService } from './bookService.ts';
+import cors from 'cors';
+const app = express();
+const port = 8881;
+
+app.use(cors())
+app.use(express.json());
+app.post('/genezio', genezioAdapters.createExpressRouter([BookService]));
+
+app.listen(port, () => {
+    console.log(`Example app listening at http://localhost:${port}`);
+});
+```
+
+Excellent! But how do we call `BookService.getAllBooks` from our clients? This is where the genezio CLI comes into the picture. We will illustrate how to convert the TypeScript client to genezio. The process for Flutter is very similar. Run the following command in the `server/` folder to generate the SDK, which you can then use to invoke the backend methods:
+
+```bash
+genezio sdk --tarball --source local --packageName genezio-sdk  --output ../client-ts/sdk --language ts --url http://localhost:8881/genezio
+```
+
+If you're curious, take a look at what was generated. It’s a tarball NPM package, which means it can be installed locally just like any other NPM package. It could also be published to a public registry such as the NPM Registry, or to a private registry. You can also export the SDK as raw files by omitting the --tarball option. This allows you to use npm link if you wish, or to add a README, or modify the package.json before publishing it.
+
+For the sake of this tutorial, we are going to install the SDK locally. In `./client-ts` folder, we can run `npm install ./sdk/genezio-sdk.tar.gz` which will install the SDK in your `node_modules` folder. 
+
+```bash
+npm install ./sdk/genezio-sdk.tar.gz
+```
```

Comment:
For flutter I don't have to install/unarchive a tarball, I directly have the sdk folder. I think that's worth mentioning.